File: attocode/src/tricks/serialization-diversity.ts
/**
 * Trick T: Serialization Diversity
 *
 * Introduces controlled variation in how context is serialized to prevent
 * few-shot pattern collapse and over-fitting to specific formats.
 *
 * Problem: When tool results and context are always serialized identically,
 * the model may develop rigid patterns - expecting certain fields in certain
 * orders, or failing when encountering slight variations.
 *
 * Solution: Introduce controlled diversity in serialization:
 * - Vary key ordering
 * - Use different quote styles
 * - Alternate between compact and pretty printing
 * - Randomize array orderings where semantically equivalent
 *
 * This helps the model stay robust to format variations.
 *
 * @example
 * ```typescript
 * import { createDiverseSerializer, serializeWithVariation } from './serialization-diversity';
 *
 * const serializer = createDiverseSerializer({
 *   variationLevel: 0.3,  // 30% variation
 *   preserveSemantics: true,
 * });
 *
 * // Each call may produce slightly different (but equivalent) output
 * const json1 = serializer.serialize(data);
 * const json2 = serializer.serialize(data);
 * // json1 and json2 are semantically equivalent but may differ in format
 * ```
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Configuration for serialization diversity.
 */
export interface DiverseSerializerConfig {
  /** Level of variation (0-1). Higher = more variation. Default: 0.3 */
  variationLevel?: number;

  /** Whether to preserve semantic meaning (never break validity) */
  preserveSemantics?: boolean;

  /** Whether to vary key ordering */
  varyKeyOrder?: boolean;

  /** Whether to vary indentation */
  varyIndentation?: boolean;

  /** Whether to vary spacing */
  varySpacing?: boolean;

  /** Whether to sometimes omit null/undefined values */
  omitNullish?: boolean;

  /** Whether to vary array element formatting */
  varyArrayFormat?: boolean;

  /** Seed for deterministic variation (useful for testing) */
  seed?: number;
}

/**
 * Serialization style options.
 */
export interface SerializationStyle {
  /** Indentation: 0 (compact), 2, 4, or 'tab' */
  indent: number | 'tab';

  /** Whether to sort keys */
  sortKeys: boolean;

  /** Key sort order: 'asc', 'desc', or 'random' */
  keySortOrder: 'asc' | 'desc' | 'random';

  /** Whether to include trailing commas (for template literals) */
  trailingComma: boolean;

  /** Space after colons */
  spaceAfterColon: boolean;

  /** Space inside brackets/braces */
  spaceInsideBrackets: boolean;

  /** Whether to omit null values */
  omitNull: boolean;

  /** Whether to omit undefined values */
  omitUndefined: boolean;

  /** Array style: 'compact', 'expanded', 'mixed' */
  arrayStyle: 'compact' | 'expanded' | 'mixed';
}

/**
 * Statistics about serialization diversity.
 */
export interface DiversityStats {
  /** Total serializations performed */
  totalSerializations: number;

  /** Distribution of styles used */
  styleDistribution: Map<string, number>;

  /** Average variation score */
  averageVariation: number;
}

/**
 * Events emitted by diverse serializer.
 */
export type SerializerEvent =
  | { type: 'serialization.performed'; style: Partial<SerializationStyle>; variation: number }
  | { type: 'style.changed'; newStyle: Partial<SerializationStyle> };

export type SerializerEventListener = (event: SerializerEvent) => void;

// =============================================================================
// RANDOM UTILITIES
// =============================================================================

/**
 * Simple seeded random number generator.
 */
class SeededRandom {
  private seed: number;

  constructor(seed?: number) {
    this.seed = seed ?? Date.now();
  }

  /**
   * Get next random number between 0 and 1.
   */
  next(): number {
    this.seed = (this.seed * 1103515245 + 12345) & 0x7fffffff;
    return this.seed / 0x7fffffff;
  }

  /**
   * Get random integer in range [min, max].
   */
  nextInt(min: number, max: number): number {
    return Math.floor(this.next() * (max - min + 1)) + min;
  }

  /**
   * Get random boolean with given probability.
   */
  nextBool(probability: number = 0.5): boolean {
    return this.next() < probability;
  }

  /**
   * Shuffle array in place.
   */
  shuffle<T>(array: T[]): T[] {
    for (let i = array.length - 1; i > 0; i--) {
      const j = this.nextInt(0, i);
      [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
  }

  /**
   * Pick random element from array.
   */
  pick<T>(array: T[]): T {
    return array[this.nextInt(0, array.length - 1)];
  }
}

// =============================================================================
// DIVERSE SERIALIZER
// =============================================================================

/**
 * Serializer that introduces controlled variation.
 */
export class DiverseSerializer {
  private config: Required<DiverseSerializerConfig>;
  private random: SeededRandom;
  private stats = {
    totalSerializations: 0,
    styleDistribution: new Map<string, number>(),
  };
  private listeners: SerializerEventListener[] = [];

  constructor(config: DiverseSerializerConfig = {}) {
    this.config = {
      variationLevel: config.variationLevel ?? 0.3,
      preserveSemantics: config.preserveSemantics ?? true,
      varyKeyOrder: config.varyKeyOrder ?? true,
      varyIndentation: config.varyIndentation ?? true,
      varySpacing: config.varySpacing ?? true,
      omitNullish: config.omitNullish ?? true,
      varyArrayFormat: config.varyArrayFormat ?? true,
      seed: config.seed ?? Date.now(),
    };
    this.random = new SeededRandom(this.config.seed);
  }

  /**
   * Serialize data with controlled variation.
   */
  serialize(data: unknown): string {
    const style = this.generateStyle();
    const result = this.serializeWithStyle(data, style);

    this.stats.totalSerializations++;
    const styleKey = this.getStyleKey(style);
    this.stats.styleDistribution.set(
      styleKey,
      (this.stats.styleDistribution.get(styleKey) || 0) + 1,
    );

    const variation = this.calculateVariation(style);
    this.emit({ type: 'serialization.performed', style, variation });

    return result;
  }

  /**
   * Serialize with a specific style.
   */
  serializeWithStyle(data: unknown, style: Partial<SerializationStyle>): string {
    const fullStyle = this.mergeWithDefaults(style);
    return this.doSerialize(data, fullStyle, 0);
  }

  /**
   * Generate a random style based on variation level.
   */
  generateStyle(): Partial<SerializationStyle> {
    const level = this.config.variationLevel;
    const style: Partial<SerializationStyle> = {};

    // Indentation
    if (this.config.varyIndentation && this.random.nextBool(level)) {
      style.indent = this.random.pick([0, 2, 4, 'tab']);
    }

    // Key ordering
    if (this.config.varyKeyOrder && this.random.nextBool(level)) {
      style.sortKeys = this.random.nextBool(0.5);
      style.keySortOrder = this.random.pick(['asc', 'desc', 'random']);
    }

    // Spacing
    if (this.config.varySpacing && this.random.nextBool(level)) {
      style.spaceAfterColon = this.random.nextBool(0.8);
      style.spaceInsideBrackets = this.random.nextBool(0.2);
    }

    // Null handling
    if (this.config.omitNullish && this.random.nextBool(level * 0.5)) {
      style.omitNull = this.random.nextBool(0.5);
      style.omitUndefined = true; // Always omit undefined in JSON
    }

    // Array formatting
    if (this.config.varyArrayFormat && this.random.nextBool(level)) {
      style.arrayStyle = this.random.pick(['compact', 'expanded', 'mixed']);
    }

    return style;
  }

  /**
   * Get a consistent style for deterministic output.
   */
  getConsistentStyle(): SerializationStyle {
    return {
      indent: 2,
      sortKeys: true,
      keySortOrder: 'asc',
      trailingComma: false,
      spaceAfterColon: true,
      spaceInsideBrackets: false,
      omitNull: false,
      omitUndefined: true,
      arrayStyle: 'expanded',
    };
  }

  /**
   * Get serialization statistics.
   */
  getStats(): DiversityStats {
    const variations: number[] = [];
    for (const [styleKey, count] of this.stats.styleDistribution) {
      const style = this.parseStyleKey(styleKey);
      variations.push(...Array(count).fill(this.calculateVariation(style)));
    }

    return {
      totalSerializations: this.stats.totalSerializations,
      styleDistribution: new Map(this.stats.styleDistribution),
      averageVariation:
        variations.length > 0 ? variations.reduce((a, b) => a + b, 0) / variations.length : 0,
    };
  }

  /**
   * Reset statistics.
   */
  resetStats(): void {
    this.stats.totalSerializations = 0;
    this.stats.styleDistribution.clear();
  }

  /**
   * Update variation level.
   */
  setVariationLevel(level: number): void {
    this.config.variationLevel = Math.max(0, Math.min(1, level));
  }

  /**
   * Subscribe to events.
   */
  on(listener: SerializerEventListener): () => void {
    this.listeners.push(listener);
    return () => {
      const idx = this.listeners.indexOf(listener);
      if (idx >= 0) this.listeners.splice(idx, 1);
    };
  }

  // Internal methods

  private mergeWithDefaults(style: Partial<SerializationStyle>): SerializationStyle {
    return {
      indent: style.indent ?? 2,
      sortKeys: style.sortKeys ?? true,
      keySortOrder: style.keySortOrder ?? 'asc',
      trailingComma: style.trailingComma ?? false,
      spaceAfterColon: style.spaceAfterColon ?? true,
      spaceInsideBrackets: style.spaceInsideBrackets ?? false,
      omitNull: style.omitNull ?? false,
      omitUndefined: style.omitUndefined ?? true,
      arrayStyle: style.arrayStyle ?? 'expanded',
    };
  }

  private doSerialize(data: unknown, style: SerializationStyle, depth: number): string {
    if (data === null) {
      return style.omitNull ? '' : 'null';
    }

    if (data === undefined) {
      return ''; // Always omit undefined in JSON
    }

    if (typeof data === 'string') {
      return JSON.stringify(data);
    }

    if (typeof data === 'number' || typeof data === 'boolean') {
      return String(data);
    }

    if (Array.isArray(data)) {
      return this.serializeArray(data, style, depth);
    }

    if (typeof data === 'object') {
      return this.serializeObject(data as Record<string, unknown>, style, depth);
    }

    return JSON.stringify(data);
  }

  private serializeArray(arr: unknown[], style: SerializationStyle, depth: number): string {
    if (arr.length === 0) {
      return style.spaceInsideBrackets ? '[ ]' : '[]';
    }

    const indent = this.getIndentString(style.indent);
    const newline = style.indent === 0 ? '' : '\n';
    const currentIndent = indent.repeat(depth);
    const itemIndent = indent.repeat(depth + 1);

    // Determine if compact
    const useCompact =
      style.arrayStyle === 'compact' ||
      (style.arrayStyle === 'mixed' &&
        arr.length <= 3 &&
        arr.every((item) => typeof item !== 'object' || item === null));

    if (useCompact || style.indent === 0) {
      const items = arr.map((item) => this.doSerialize(item, style, depth + 1)).filter(Boolean);
      const inner = items.join(', ');
      return style.spaceInsideBrackets ? `[ ${inner} ]` : `[${inner}]`;
    }

    const items = arr
      .map((item) => this.doSerialize(item, style, depth + 1))
      .filter(Boolean)
      .map((item) => `${itemIndent}${item}`);

    return `[${newline}${items.join(`,${newline}`)}${newline}${currentIndent}]`;
  }

  private serializeObject(
    obj: Record<string, unknown>,
    style: SerializationStyle,
    depth: number,
  ): string {
    let keys = Object.keys(obj);

    // Filter keys based on style
    keys = keys.filter((key) => {
      const value = obj[key];
      if (value === undefined) return false;
      if (value === null && style.omitNull) return false;
      return true;
    });

    if (keys.length === 0) {
      return style.spaceInsideBrackets ? '{ }' : '{}';
    }

    // Sort keys based on style
    if (style.sortKeys) {
      if (style.keySortOrder === 'random') {
        this.random.shuffle(keys);
      } else {
        keys.sort();
        if (style.keySortOrder === 'desc') {
          keys.reverse();
        }
      }
    }

    const indent = this.getIndentString(style.indent);
    const newline = style.indent === 0 ? '' : '\n';
    const currentIndent = indent.repeat(depth);
    const propIndent = indent.repeat(depth + 1);
    const colonSpace = style.spaceAfterColon ? ' ' : '';

    if (style.indent === 0) {
      const pairs = keys.map((key) => {
        const value = this.doSerialize(obj[key], style, depth + 1);
        return `${JSON.stringify(key)}:${colonSpace}${value}`;
      });
      return `{${pairs.join(',')}}`;
    }

    const pairs = keys.map((key) => {
      const value = this.doSerialize(obj[key], style, depth + 1);
      return `${propIndent}${JSON.stringify(key)}:${colonSpace}${value}`;
    });

    return `{${newline}${pairs.join(`,${newline}`)}${newline}${currentIndent}}`;
  }

  private getIndentString(indent: number | 'tab'): string {
    if (indent === 'tab') return '\t';
    return ' '.repeat(indent);
  }

  private getStyleKey(style: Partial<SerializationStyle>): string {
    return `${style.indent ?? 2}-${style.sortKeys ?? true}-${style.keySortOrder ?? 'asc'}`;
  }

  private parseStyleKey(key: string): Partial<SerializationStyle> {
    const [indent, sortKeys, keySortOrder] = key.split('-');
    return {
      indent: indent === 'tab' ? 'tab' : parseInt(indent, 10),
      sortKeys: sortKeys === 'true',
      keySortOrder: keySortOrder as 'asc' | 'desc' | 'random',
    };
  }

  private calculateVariation(style: Partial<SerializationStyle>): number {
    let variation = 0;
    const defaults = this.getConsistentStyle();

    if (style.indent !== undefined && style.indent !== defaults.indent) variation += 0.2;
    if (style.sortKeys !== undefined && style.sortKeys !== defaults.sortKeys) variation += 0.2;
    if (style.keySortOrder !== undefined && style.keySortOrder !== defaults.keySortOrder)
      variation += 0.2;
    if (style.spaceAfterColon !== undefined && style.spaceAfterColon !== defaults.spaceAfterColon)
      variation += 0.1;
    if (style.omitNull !== undefined && style.omitNull !== defaults.omitNull) variation += 0.1;
    if (style.arrayStyle !== undefined && style.arrayStyle !== defaults.arrayStyle)
      variation += 0.2;

    return Math.min(1, variation);
  }

  private emit(event: SerializerEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a diverse serializer.
 *
 * @example
 * ```typescript
 * const serializer = createDiverseSerializer({
 *   variationLevel: 0.3,
 *   preserveSemantics: true,
 * });
 *
 * // Serialize tool results with variation
 * const result1 = serializer.serialize({ files: ['a.ts', 'b.ts'], count: 2 });
 * const result2 = serializer.serialize({ files: ['a.ts', 'b.ts'], count: 2 });
 *
 * // Results are semantically equivalent but may differ in format:
 * // result1: {"count": 2, "files": ["a.ts", "b.ts"]}
 * // result2: {"files": ["a.ts", "b.ts"], "count": 2}
 * ```
 */
export function createDiverseSerializer(config: DiverseSerializerConfig = {}): DiverseSerializer {
  return new DiverseSerializer(config);
}

// =============================================================================
// UTILITIES
// =============================================================================

/**
 * Quick helper to serialize with variation.
 */
export function serializeWithVariation(data: unknown, variationLevel: number = 0.3): string {
  const serializer = createDiverseSerializer({ variationLevel });
  return serializer.serialize(data);
}

/**
 * Generate multiple serializations of the same data.
 * Useful for training data or testing robustness.
 */
export function generateVariations(
  data: unknown,
  count: number,
  variationLevel: number = 0.5,
): string[] {
  const serializer = createDiverseSerializer({ variationLevel });
  const results: string[] = [];

  for (let i = 0; i < count; i++) {
    results.push(serializer.serialize(data));
  }

  return results;
}

/**
 * Apply diversity to tool call arguments.
 */
export function diversifyToolArgs(
  args: Record<string, unknown>,
  variationLevel: number = 0.3,
): string {
  const serializer = createDiverseSerializer({
    variationLevel,
    preserveSemantics: true,
  });
  return serializer.serialize(args);
}

/**
 * Apply diversity to tool results.
 */
export function diversifyToolResult(result: unknown, variationLevel: number = 0.3): string {
  const serializer = createDiverseSerializer({
    variationLevel,
    preserveSemantics: true,
    varyArrayFormat: true,
  });
  return serializer.serialize(result);
}

/**
 * Format diversity stats for display.
 */
export function formatDiversityStats(stats: DiversityStats): string {
  const lines = [
    `Serialization Diversity Statistics:`,
    `  Total serializations: ${stats.totalSerializations}`,
    `  Average variation: ${(stats.averageVariation * 100).toFixed(1)}%`,
    '',
    '  Style distribution:',
  ];

  for (const [style, count] of stats.styleDistribution) {
    const percent = ((count / stats.totalSerializations) * 100).toFixed(1);
    lines.push(`    ${style}: ${count} (${percent}%)`);
  }

  return lines.join('\n');
}

/**
 * Validate that two serializations are semantically equivalent.
 */
export function areSemanticEquivalent(json1: string, json2: string): boolean {
  try {
    const obj1 = JSON.parse(json1);
    const obj2 = JSON.parse(json2);
    return (
      JSON.stringify(obj1, Object.keys(obj1).sort()) ===
      JSON.stringify(obj2, Object.keys(obj2).sort())
    );
  } catch {
    return false;
  }
}

File: attocode/attocode_py/src/attocode/tricks/serialization_diversity.py
"""Serialization diversity for cache-busting (Trick T).

Varies JSON serialization style to prevent the LLM from
over-fitting to specific formatting patterns.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass
class SerializationStyle:
    """A serialization style configuration."""

    indent: int | None = 2
    sort_keys: bool = True
    key_sort_order: str = "asc"  # asc, desc, random
    space_after_colon: bool = True
    omit_null: bool = False
    array_style: str = "expanded"  # compact, expanded


@dataclass
class DiversityStats:
    """Statistics about serialization diversity."""

    total_serializations: int = 0
    style_distribution: dict[str, int] = field(default_factory=dict)


@dataclass
class DiverseSerializerConfig:
    """Configuration for diverse serialization."""

    variation_level: float = 0.3  # 0.0 = no variation, 1.0 = max variation
    preserve_semantics: bool = True
    vary_key_order: bool = True
    vary_indentation: bool = True
    omit_nullish: bool = True
    vary_array_format: bool = True
    seed: int | None = None


DiverseSerializerEventListener = Callable[[str, dict[str, Any]], None]


class _SeededRandom:
    """Simple seeded PRNG for deterministic variation."""

    def __init__(self, seed: int) -> None:
        self._seed = seed & 0x7FFFFFFF

    def next_float(self) -> float:
        """Get next random float in [0, 1)."""
        self._seed = (self._seed * 1103515245 + 12345) & 0x7FFFFFFF
        return self._seed / 0x7FFFFFFF

    def next_bool(self, probability: float = 0.5) -> bool:
        """Get next random boolean with given probability."""
        return self.next_float() < probability

    def next_int(self, low: int, high: int) -> int:
        """Get next random int in [low, high]."""
        return low + int(self.next_float() * (high - low + 1))


class DiverseSerializer:
    """Serializes data with controlled style variation.

    Varies indentation, key ordering, spacing, and array formatting
    while preserving semantic equivalence.
    """

    def __init__(self, config: DiverseSerializerConfig | None = None) -> None:
        self._config = config or DiverseSerializerConfig()
        import time
        seed = self._config.seed if self._config.seed is not None else int(time.time() * 1000)
        self._rng = _SeededRandom(seed)
        self._stats = DiversityStats()
        self._listeners: list[DiverseSerializerEventListener] = []

    def serialize(self, data: Any) -> str:
        """Serialize with a randomly generated style."""
        style = self.generate_style()
        result = self.serialize_with_style(data, style)
        self._stats.total_serializations += 1

        # Track style distribution
        style_key = f"{style.indent}-{style.sort_keys}-{style.key_sort_order}"
        self._stats.style_distribution[style_key] = (
            self._stats.style_distribution.get(style_key, 0) + 1
        )

        self._emit("serialization.performed", {"style": style_key})
        return result

    def serialize_with_style(self, data: Any, style: SerializationStyle) -> str:
        """Serialize data with a specific style."""
        # Process the data (omit nulls, sort keys, etc.)
        processed = self._process_value(data, style)

        indent = style.indent
        separators = None
        if not style.space_after_colon:
            separators = (",", ":")
        elif indent is None:
            separators = (", ", ": ")

        return json.dumps(
            processed,
            indent=indent,
            sort_keys=False,  # We handle sorting ourselves
            ensure_ascii=False,
            separators=separators,
        )

    def generate_style(self) -> SerializationStyle:
        """Generate a randomized style based on variation level."""
        level = self._config.variation_level
        style = SerializationStyle()

        if self._config.vary_indentation and self._rng.next_bool(level):
            style.indent = self._rng.next_int(0, 4)
            if style.indent == 0:
                style.indent = None  # compact

        if self._config.vary_key_order and self._rng.next_bool(level):
            choices = ["asc", "desc", "random"]
            idx = self._rng.next_int(0, 2)
            style.key_sort_order = choices[idx]
            style.sort_keys = style.key_sort_order != "random"

        if self._rng.next_bool(level * 0.5):
            style.space_after_colon = not style.space_after_colon

        if self._config.omit_nullish and self._rng.next_bool(level):
            style.omit_null = True

        return style

    def get_consistent_style(self) -> SerializationStyle:
        """Get a deterministic baseline style."""
        return SerializationStyle(
            indent=2,
            sort_keys=True,
            key_sort_order="asc",
            space_after_colon=True,
            omit_null=False,
            array_style="expanded",
        )

    def get_stats(self) -> DiversityStats:
        """Get diversity statistics."""
        return DiversityStats(
            total_serializations=self._stats.total_serializations,
            style_distribution=dict(self._stats.style_distribution),
        )

    def reset_stats(self) -> None:
        """Reset statistics."""
        self._stats = DiversityStats()

    def set_variation_level(self, level: float) -> None:
        """Set the variation level (0.0 to 1.0)."""
        self._config.variation_level = max(0.0, min(1.0, level))

    def on(self, listener: DiverseSerializerEventListener) -> Callable[[], None]:
        """Subscribe to serializer events."""
        self._listeners.append(listener)

        def unsubscribe() -> None:
            try:
                self._listeners.remove(listener)
            except ValueError:
                pass

        return unsubscribe

    def _process_value(self, value: Any, style: SerializationStyle) -> Any:
        """Recursively process a value according to the style."""
        if isinstance(value, dict):
            result: dict[str, Any] = {}
            keys = list(value.keys())

            if style.key_sort_order == "asc":
                keys.sort()
            elif style.key_sort_order == "desc":
                keys.sort(reverse=True)
            elif style.key_sort_order == "random":
                # Shuffle with seeded RNG
                for i in range(len(keys) - 1, 0, -1):
                    j = self._rng.next_int(0, i)
                    keys[i], keys[j] = keys[j], keys[i]

            for k in keys:
                v = value[k]
                if style.omit_null and v is None:
                    continue
                result[k] = self._process_value(v, style)
            return result
        elif isinstance(value, list):
            return [self._process_value(item, style) for item in value]
        else:
            return value

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for listener in self._listeners:
            try:
                listener(event, data)
            except Exception:
                pass


def serialize_with_variation(data: Any, variation_level: float = 0.3) -> str:
    """One-shot diverse serialization."""
    serializer = DiverseSerializer(DiverseSerializerConfig(variation_level=variation_level))
    return serializer.serialize(data)


def generate_variations(data: Any, count: int, variation_level: float = 0.5) -> list[str]:
    """Generate multiple different serializations of the same data."""
    serializer = DiverseSerializer(DiverseSerializerConfig(variation_level=variation_level))
    return [serializer.serialize(data) for _ in range(count)]


def are_semantic_equivalent(json1: str, json2: str) -> bool:
    """Check if two JSON strings are semantically equivalent."""
    try:
        obj1 = json.loads(json1)
        obj2 = json.loads(json2)
        # Normalize by re-serializing with sorted keys
        norm1 = json.dumps(obj1, sort_keys=True)
        norm2 = json.dumps(obj2, sort_keys=True)
        return norm1 == norm2
    except (json.JSONDecodeError, TypeError):
        return False


def find_most_compact(data: Any, max_attempts: int = 10) -> tuple[str, SerializationStyle]:
    """Find the most compact serialization for data.

    Tries multiple styles and returns the shortest output.
    Useful for minimizing token usage in constrained contexts.
    """
    serializer = DiverseSerializer(DiverseSerializerConfig(
        variation_level=0.8,
        vary_indentation=True,
        vary_key_order=True,
        vary_array_format=True,
    ))

    best_result = ""
    best_style = serializer.get_consistent_style()
    best_len = float("inf")

    # Always try the compact baseline
    compact_style = SerializationStyle(
        indent=None,
        sort_keys=True,
        key_sort_order="asc",
        space_after_colon=False,
        omit_null=True,
        array_style="compact",
    )
    compact_result = serializer.serialize_with_style(data, compact_style)
    if len(compact_result) < best_len:
        best_result = compact_result
        best_style = compact_style
        best_len = len(compact_result)

    # Try random variations
    for _ in range(max_attempts):
        style = serializer.generate_style()
        result = serializer.serialize_with_style(data, style)
        if len(result) < best_len:
            best_result = result
            best_style = style
            best_len = len(result)

    return best_result, best_style


def measure_compression_ratio(data: Any) -> dict[str, Any]:
    """Measure the compression ratio between different serialization styles."""
    # Baseline: pretty-printed
    pretty = json.dumps(data, indent=2, sort_keys=True)
    # Compact: no whitespace
    compact = json.dumps(data, separators=(",", ":"), sort_keys=True)

    # Diverse
    serializer = DiverseSerializer(DiverseSerializerConfig(variation_level=0.5))
    diverse = serializer.serialize(data)

    return {
        "pretty_length": len(pretty),
        "compact_length": len(compact),
        "diverse_length": len(diverse),
        "compact_ratio": len(compact) / max(1, len(pretty)),
        "diverse_ratio": len(diverse) / max(1, len(pretty)),
        "savings_chars": len(pretty) - len(compact),
        "savings_tokens_approx": (len(pretty) - len(compact)) // 4,
    }


class AdaptiveSerializer(DiverseSerializer):
    """Serializer that adapts variation level based on context.

    Increases variation when the LLM seems to be repeating patterns,
    decreases it when the output is already diverse.
    """

    def __init__(self, config: DiverseSerializerConfig | None = None) -> None:
        super().__init__(config)
        self._recent_styles: list[str] = []
        self._max_style_history: int = 20

    def serialize(self, data: Any) -> str:
        """Serialize with adaptive variation."""
        # Check if we're in a repetition rut
        if self._is_style_repetitive():
            # Temporarily boost variation
            saved = self._config.variation_level
            self._config.variation_level = min(1.0, saved + 0.3)
            result = super().serialize(data)
            self._config.variation_level = saved
        else:
            result = super().serialize(data)

        # Track the style used
        style = f"{self._stats.total_serializations}"
        self._recent_styles.append(style)
        if len(self._recent_styles) > self._max_style_history:
            self._recent_styles.pop(0)

        return result

    def _is_style_repetitive(self) -> bool:
        """Check if recent styles are too similar."""
        if len(self._recent_styles) < 5:
            return False
        # Check if the last 5 styles are all the same
        last_5 = self._recent_styles[-5:]
        unique = len(set(last_5))
        return unique <= 2
